### ISSUE TITLE
Preserve user settings in universal variables

### ISSUE THREAD

**Issue by andreiborisov** [gazorby/fish-exa#9]
1. I think it's better to skip erasing user settings in universal variables, as it's a expected behaviour for the most software: https://github.com/gazorby/fish-exa/blob/19cc781c2f4e332093f0bf5e870a0e04ba57e08c/conf.d/fish-exa.fish#L69

We should only erase private variables prefixed with `_`.
What do you think?

2. Another thing, it's probably better to check if a user setting in universal variable exist before rewriting it: https://github.com/andreiborisov/sponge/blob/main/conf.d/sponge.fish Otherwise, you can possibly write on top of existing user configuration (for example, when the user transferred universal variables via dotfiles and setting up a new machine, variables will exist before the plugin installation in that case).

**Comment by andreiborisov** [comment]
I had the same setup on my own plugin and was surprised by its behaviour😅 https://github.com/andreiborisov/sponge/commit/2ff7d9941b2095311771bb29942958e571e7080e

**Comment by gazorby** [comment]
Hi @andreiborisov!

> I think it's better to skip erasing user settings in universal variables

The main reason user settings are deleted on uninstall is to allow reseting the plugin to a clean state, from which you can start tweaking settings again if something went wrong. Also, why would user settings from an absent plugin would be useful if you _explictly uninstalled it_?

One the other hand, i agree that persisting them after plugin _update_ should be the expected behaviour, which is not the case right now (and it might be the reason you opened this, sorry for that if it's the case :/):

https://github.com/gazorby/fish-exa/blob/19cc781c2f4e332093f0bf5e870a0e04ba57e08c/conf.d/fish-exa.fish#L64

Moreover, missing documentation about this doesn't help.

> it's probably better to check if a user setting in universal variable exist before rewriting it

Sure. An option would be to use global scoped variables for default configuration (when no user configuration exists). This way users can use universal variables to override defaults and they would persist after plugin removal.

**Comment by andreiborisov** [comment]
> One the other hand, i agree that persisting them after plugin update should be the expected behaviour, which is not the case right now.

Yes, I believe Fisher fires `uninstall` and `install` events on update and there is no way to distinguish an update from a manual uninstall.

> An option would be to use global scoped variables for default configuration (when no user configuration exists). This way users can use universal variables to override defaults and they would persist after plugin removal.

That what we were using in pure prompt, but at the end we've decided that it's not the best approach because:

1. It's often being confusing for the user that they can use universal variables, but cannot use global variables. User might wanna use different settings for different shell sessions (so have both global and universal variables configured).
2. If you'll check for existence of a universal variable in `conf.d/fish-exa.fish` and set a global variable in case there is no a universal one, the user will be required to use `set` with a `-U` flag to set their value, which they often forget (see 1). But even more important, it will require shell restart to take effect, which goes against description of universal variables in fish docs (instant effect across all fish sessions). So it's kinda counterintuitive.

tldr: Keeping in my mind these points, in my opinion, it's easier both from the point of implementation and the usage to check if a universal variable is set, and assign it the default value, if it's not (in `conf.d/fish-exa.fish`).

> The main reason user settings are deleted on uninstall is to allow reseting the plugin to a clean state, from which you can start tweaking settings again if something went wrong.

This is a good point, something I'm worried about too... However, I'm leaning towards pointing users to Sponge documentation, where there is a clearly defined list of used variables. Since it's the same for `fish-exa` perhaps a section with an instruction on how to clean all of the variables (`set -e EXA_*`) would be sufficient?

**Comment by gazorby** [comment]
> there is no way to distinguish an update from a manual uninstall

In fact, fisher emit a [separate event for update](https://github.com/jorgebucaran/fisher#event-system), so you can have different logic for both update and uninstall.
Currently, `__fish_exa_update` is overriden with the same behaviour as default (uninstall then install), but we could fix it by skipping uninstall and just removing aliases.

> perhaps a section with an instruction on how to clean all of the variables (set -e EXA_*) would be sufficient?

Your're right that common behaviour is to left user settings untouched, so mentioning this in the readme should be enough. Maybe reseting plugin state is a missing feature of fisher, it reminds me the reset command of the [scoop package manager (Windows)](https://github.com/ScoopInstaller/Scoop/wiki/Commands)

> User might wanna use different settings for different shell sessions (so have both global and universal variables configured)

> But even more important, it will require shell restart to take effect

Another pattern would be to just seperate "runtime configuration" from user configuration.
For example, we can use private, universal variables (starting with `_`) to store default configuration, and replace their value _at runtime_, with the equivalent public variables's values if they exists:

The plugin would comes with default settings, stored in private variables:
```fish
set -U _EXA_DEFAULT_OPTIONS "--group" "--header" "--group-directories-first"
set -U _EXA_LL_OPTIONS "--long"
```

And resolve user config at runtime:

```fish
function __fish_exa_config
    set -l options
    for a in $argv
        if set -q $a
            set -a options $$a
        else
            set a (string join "" "_" $a)
            set -a options $$a
        end
    end
    echo $options
end

alias --save "$name" "exa (string split ' ' -- (__fish_exa_config EXA_STANDARD_OPTIONS EXA_LL_OPTIONS)"
```

In addition, we could change the uninstall function to erase global variables only, this way one can decide to configure the plugin only using global vars in config.fish (let's say, for testing purposes) so uninstalling it would reset the state, but only for the current shell.

Combined with runtime configuration this approach would have the following advantages:
- User have full control over configuration scoping (can be either global, universal or a mix of both)
- Any change takes effect immediately
- Using global variables only, it's possible to reset the plugin back to its initial state (only in the current shell instance) just by reinstalling it.

What do you think?

**Comment by andreiborisov** [comment]
> In fact, fisher emit a [separate event for update](https://github.com/jorgebucaran/fisher#event-system), so you can have different logic for both update and uninstall.

Ah, so it was fixed, nice, I wasn't aware!

> Maybe reseting plugin state is a missing feature of fisher, it reminds me the reset command of the [scoop package manager (Windows)](https://github.com/ScoopInstaller/Scoop/wiki/Commands)

Knowing Jorge's love for simplicity, we shouldn't hold our breath on this one😁

> Another pattern would be to just separate "runtime configuration" from user configuration.

That's absolutely a valid option. I think it comes down to personal preferences here, since it would be behave in a similar way for the user.

One thing to note however, this approach seem to require more maintenance from the plugin author compared to simplistic "just read the variable". So it wouldn't be my first option personally.

> Using global variables only, it's possible to reset the plugin back to its initial state (only in the current shell instance) just by reinstalling it.

I'm not sure why would someone want to reset the plugin to its initial state for the current shell only. If you worried about validity of user configuration state, you can use your runtime approach to check if the configuration is valid and set the default values for invalid parameters, possibly with a message for the user.

> What do you think?

I will be happy with either approach🙌🏻 (as long as the configuration stored in universal variables is persisted, so I can throw it in my dotfiles).

**Comment by gazorby** [comment]
Hi @andreiborisov!

A bit late but user-defined variables are finally preserved as of ad73006f309e2455fa140049cc23fcf6be28f34d, enjoy customization!

**Comment by andreiborisov** [comment]
Not late at all — still using your plugin. Thank you for the update!☺️
